packages/react-utilities/src/compose/types.ts

```diff
@@ -1,5 +1,56 @@
 import * as React from 'react';
 
+/**
+ * Generic record of possible ShorthandProps
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type SlotPropsRecord = Record<string, Record<string, any> | undefined>;
+
+export type ShorthandPropsRecord<Record extends SlotPropsRecord = SlotPropsRecord> = {
+  [K in keyof Record]: ShorthandProps<NonNullable<Record[K]>>;
+};
+
+export type ObjectShorthandPropsRecord<Record extends SlotPropsRecord = SlotPropsRecord> = {
+  [K in keyof Record]: ObjectShorthandProps<NonNullable<Record[K]>>;
+};
+
+export type ShorthandRenderFunction<Props> = (Component: React.ElementType<Props>, props: Props) => React.ReactNode;
+
+export type ShorthandProps<Props = {}> =
+  | React.ReactChild
+  | React.ReactNodeArray
+  | React.ReactPortal
+  | number
+  | null
+  | undefined
+  | ObjectShorthandProps<Props>;
+
+export type ObjectShorthandProps<Props extends { children?: React.ReactNode } = {}> = Props &
+  Pick<ComponentProps, 'as'> & {
```

Comment:
https://github.com/microsoft/fluentui/blob/master/rfcs/convergence/simplify-prop-merging.md#do-not-support-as-prop-for-slots-that-render-native-dom-elements

`as` should be added only if the component itself supports it, we can remove `Pick<>` from there